src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,669 @@
+            throws InvalidAlgorithmParameterException,
+                   NoSuchAlgorithmException {
+        if (alg == null) {
+            throw new NullPointerException(
+                    "the algorithm for the SecretKey return value must not be"
+                    + " null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                    "the algorithm for the SecretKey return value must not be "
+                    + "empty");
+        }
+        Objects.requireNonNull(derivationSpec);
+        if (checkSpiNonNull(theOne)) {
+            return theOne.spi().engineDeriveKey(alg, derivationSpec);
+        } else {
+            return (SecretKey) chooseProvider(alg, derivationSpec);
+        }
+    }
+
+    /**
+     * Derives a key, returns raw data as a byte array.
+     *
+     * @param derivationSpec
+     *         the object describing the inputs to the derivation function
+     *
+     * @return the derived key in its raw bytes
+     *
+     * @throws InvalidAlgorithmParameterException
+     *         if the information contained within the {@code derivationSpec} is
+     *         invalid
+     * @throws UnsupportedOperationException
+     *         if the derived keying material is not extractable
+     * @throws NullPointerException
+     *         if {@code derivationSpec} is null
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider
+     *         Selection</a>
+     *
+     */
+    public byte[] deriveData(AlgorithmParameterSpec derivationSpec)
+            throws InvalidAlgorithmParameterException {
+
+        Objects.requireNonNull(derivationSpec);
+        if (checkSpiNonNull(theOne)) {
+            return theOne.spi().engineDeriveData(derivationSpec);
+        } else {
+            try {
+                return (byte[]) chooseProvider(null, derivationSpec);
+            } catch (NoSuchAlgorithmException e) {
+                // this will never be thrown in the deriveData case
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    /**
+     * Use the firstSpi as the chosen KDFSpi and set the fields accordingly
+     */
+    private void useFirstSpi() {
+        if (checkSpiNonNull(theOne)) return;
+
+        synchronized (lock) {
+            theOne = candidate;
+        }
+    }
+
+    /**
+     * Selects the provider which supports the passed {@code algorithm} and
+     * {@code derivationSpec} values, and assigns the global spi and provider
+     * variables if they have not been assigned yet.
+     * <p>
+     * If the spi has already been set, it will just return the result.
+     */
+    private Object chooseProvider(String algorithm,
+                                  AlgorithmParameterSpec derivationSpec)
+            throws InvalidAlgorithmParameterException,
+                   NoSuchAlgorithmException {
+
+        boolean isDeriveData = (algorithm == null);
+
+        if (checkSpiNonNull(theOne)) {
```

Comment:
This check is still needed inside the `synchronized` block below, otherwise another thread waiting on the same lock with perform DPS again after the first one exits.

In fact, since you already check for the same condition before each `chooseProvider` call, it is only necessary inside the `synchronized` block.